### ISSUE TITLE
Make `unreachable` panic in ReleaseFast, introduce explicit `@optimizeAway()` builtin 

### ISSUE THREAD

**Issue by Jarred-Sumner** [ziglang/zig#19664]
Today, the cost of an incorrect `unreachable` in ReleaseFast is, at best, a segfault. At worst, it's the start of a security vulnerability.

The developer's intent when using `unreachable` is usually to assert "This should never happen". Outside of comptime, the developer rarely means "Compiler, optimize this code in a way that leverages the fact this code will never be reached". This distinction is important and the keyword `unreachable` does not make that clear enough to the developer.

This proposal suggests `unreachable` becomes a panic in `ReleaseFast`, and instead for something like `@optimizeAway()` or `@unreachable()` to be used for compiler optimizations. It should be extremely clear to the developer when the compiler will use this information in optimizations.

#### Just use `@panic()`?

Developers reach for the easiest thing, and `unreachable` is easier than `@panic(msg)`.

In Zig's codebase, there are nearly 2x more files using unreachable in a switch compared to `@panic` being used anywhere at all:

```zig
❯ rg " => unreachable" -l -t zig | wc -l
268

❯ rg " @panic\(" -l -t zig | wc -l
135
```

If we search just for the string ` unreachable` (even though that is a more noisy number):
```zig
❯ rg " unreachable" -l -t zig | wc -l
536
```

### Recent incorrect usages of `unreachable` in the Zig compiler 

Running the following command in the Zig repository shows recent cases where an `unreachable` appears in a commit diff:
```sh
git log -p | rg -A10 -B10 '^-.* unreachable'
```

Appears that the URl was not always already validated, or req.uri.port was sometimes missing:
```diff
-            // URI has already been validated so this cannot fail.
-            const default_port = (uriProtocol(req.uri) catch unreachable).port();
             try w.writeAll("host: ");
-            try req.uri.writeToStream(.{
-                .authority = true,
-                .port = req.uri.port.? != default_port,
-            }, w);
+            try req.uri.writeToStream(.{ .authority = true }, w);
```

`uri.user` and/or `uri.password` can exceed the max lengths:
```diff
     pub fn value(uri: Uri, out: []u8) []u8 {
-        assert(uri.user == null or uri.user.?.len <= max_user_len);
-        assert(uri.password == null or uri.password.?.len <= max_password_len);
-
-        @memcpy(out[0..prefix.len], prefix);
-
         var buf: [max_user_len + ":".len + max_password_len]u8 = undefined;
-        const unencoded = std.fmt.bufPrint(&buf, "{s}:{s}", .{
-            uri.user orelse "", uri.password orelse "",
-        }) catch unreachable;
-        const base64 = std.base64.standard.Encoder.encode(out[prefix.len..], unencoded);
+        var stream = std.io.fixedBufferStream(&buf);
+        stream.writer().print("{user}", .{uri.user orelse Uri.Component.empty}) catch
+            unreachable;
+        assert(stream.pos <= max_user_len);
+        stream.writer().print(":{password}", .{uri.password orelse Uri.Component.empty}) catch
+            unreachable;
```

Seems that `tag == .c` is another case where the branch can be reached?
```diff
@@ -208,25 +203,17 @@ pub const File = struct {
         emit: Compilation.Emit,
         options: OpenOptions,
     ) !*File {
-        const tag = Tag.fromObjectFormat(comp.root_mod.resolved_target.result.ofmt);
-        switch (tag) {
-            .c => {
-                const ptr = try C.createEmpty(arena, comp, emit, options);
-                return &ptr.base;
-            },
-            inline else => |t| {
-                if (build_options.only_c) unreachable;
-                const ptr = try t.Type().createEmpty(arena, comp, emit, options);
+        switch (Tag.fromObjectFormat(comp.root_mod.resolved_target.result.ofmt)) {
+            inline else => |tag| {
+                if (tag != .c and build_options.only_c) unreachable;
+                const ptr = try tag.Type().createEmpty(arena, comp, emit, options);
                 return &ptr.base;
             },
         }
     }
```

Looks like the `error.FileLocksNotSupported` is not unreachable:
```diff
     const fd = posix.openatZ(self.fd, pathname, flags, 0) catch |err| switch (err) {
-        error.FileLocksNotSupported => unreachable,
+        error.FileLocksNotSupported => return error.Unexpected,
         else => |e| return e,
```

These errors are not unreachable:
```diff
-                error.PathAlreadyExists => unreachable, // not possible for "NUL"
-                error.PipeBusy => unreachable, // not possible for "NUL"
-                error.FileNotFound => unreachable, // not possible for "NUL"
-                error.AccessDenied => unreachable, // not possible for "NUL"
-                error.NameTooLong => unreachable, // not possible for "NUL"
-                error.WouldBlock => unreachable, // not possible for "NUL"
-                error.NetworkNotFound => unreachable, // not possible for "NUL"
+                error.PathAlreadyExists => return error.Unexpected, // not possible for "NUL"
+                error.PipeBusy => return error.Unexpected, // not possible for "NUL"
+                error.FileNotFound => return error.Unexpected, // not possible for "NUL"
+                error.AccessDenied => return error.Unexpected, // not possible for "NUL"
+                error.NameTooLong => return error.Unexpected, // not possible for "NUL"
+                error.WouldBlock => return error.Unexpected, // not possible for "NUL"
+                error.NetworkNotFound => return error.Unexpected, // not possible for "NUL"
+                error.AntivirusInterference => return error.Unexpected, // not possible for "NUL"
```

`error.AccessDenied` is not unreachable:
```diff
-                error.AccessDenied => unreachable,
+                error.AccessDenied => return error.Unexpected,
                 else => |e| return e,
```
Another occurrence of the same:
```diff
-        error.AccessDenied => unreachable,
+        error.AccessDenied => return error.Unexpected,
```

Corrupt stacktraces happen:

```diff
-    const field_index = sema.structFieldIndex(block, stack_trace_ty, field_name, src) catch |err| switch (err) {
-        error.NeededSourceLocation, error.GenericPoison, error.ComptimeReturn, error.ComptimeBreak => unreachable,
-        else => |e| return e,
+    const field_index = sema.structFieldIndex(block, stack_trace_ty, field_name, .unneeded) catch |err| switch (err) {
+        error.AnalysisFail, error.NeededSourceLocation => @panic("std.builtin.StackTrace is corrupt"),
+        error.GenericPoison, error.ComptimeReturn, error.ComptimeBreak => unreachable,
```

I'm confident there are many more cases where this has happened. In Bun's codebase, we have banned `std.debug.assert` since there have been many cases where unreachable code was in fact not unreachable. Note that `std.debug.assert` is implemented as `if (!condition) unreachable`

**Comment by silversquirl** [comment]
> It should be extremely clear to the developer when the compiler will use this information in optimizations.

I would argue the name `unreachable` is *very* clear - it means "I guarantee this line will never be reached". If people are using it without understanding what it does, that's a documentation problem.

**Comment by xdBronch** [comment]
theres also precedent for `unreachable` to mean this or something very similar, C has `__builtin_unreachable()` and rust has `std::hint::unreachable_unchecked()` which is functionally identical to zigs. i think the solution to this is stricter testing and/or using ReleaseSafe which is [already planned for zig](https://github.com/ziglang/zig/issues/15194)

**Comment by Jarred-Sumner** [comment]
> I would argue the name unreachable is very clear - it means "I guarantee this line will never be reached". If people are using it without understanding what it does, that's a documentation problem.

There are enough commits in Zig's codebase where developers made this mistake to disprove this

> theres also precedent for `unreachable` to mean this or something very similar, C has `__builtin_unreachable()` and rust has `std::hint::unreachable_unchecked()` which is functionally identical to zigs. i think the solution to this is stricter testing and/or using ReleaseSafe which is [already planned for zig](https://github.com/ziglang/zig/issues/15194)

In both of those cases, `unreachable` is not a keyword. `__builtin_unreachable()` is a scary-sounding function. rust's `std::hint__unreachable_unchecked()` makes it extremely clear that it is **unchecked**, which is why ` unreachable!` is the default people use instead (which panics)

**Comment by Vexu** [comment]
Another solution to your use case would be implementing #13339 and compiling Bun with `-OReleaseFast+unreachable`.

**Comment by xdBronch** [comment]
> __builtin_unreachable() is a scary-sounding function

i dont really understand your point? i dont see how it being a function instead of a keyword all of a sudden makes it scary
> rust's std::hint__unreachable_unchecked() makes it extremely clear that it is unchecked

thats fair but surely we shouldnt shape the language around things that people who are unfamiliar with zig assume about it? zigs unreachable is actually the safest of the 3 here, Cs is completely unchecked and rusts is unchecked in any release mode afaik, zig gives you the choice to keep these things checked while still getting optimizations

**Comment by mlugg** [comment]
> Outside of comptime, the developer rarely means "Compiler, optimize this code in a way that leverages the fact that this code will never be reached."

On the contrary, this is *precisely* what I mean when I write an assertion. When I write `assert(xyz)`, I am stating that `xyz` will always be true. If it is not true in any case, I of course would like to know about that when safety is enabled; but otherwise, I expect the optimizer to utilize that fact when it makes sense. It's no different than any other safety check.

This proposal doesn't just lose the optimizations made possible by certain code paths being `unreachable` (although that is still itself a major loss); it also requires unnecessary code to be evaluated, even in `ReleaseFast`. It is not uncommon for me to write code like `assert(nonTrivialFunction(foo))` -- I absolutely **do not want that function to be called** in unsafe optimized builds, because it wastes runtime to check for a bug which I ought to have already caught in debugging. I highly doubt you would propose to make *other* safety checks apply in `ReleaseFast`; why should `unreachable` be any different?

> At worst, it's the start of a security vulnerability.

**If you are not confident in the correctness of your assertions, you should not run the code in security-critical environments.** Assertions are intended to be factual statements about how your code's logic works: an assertion failing indicates a bug which, in all likelihood, could manifest regardless of the assertion. We can't magically guard against incorrect code. It is not appropriate to complicate the language, and hugely limit optimization capabilities, just in order for certain incorrect code to manifest itself in a safer way in release builds.

If you are working in a complex codebase where performance is less critical and there are many trivial assertions, it may indeed be appropriate to create a "custom" assertion function which panics in all release modes. This should not be the default.

> [...] the keyword `unreachable` does not make that clear enough to the developer.

As already mentioned by the commenters above me, the keyword is completely unambiguous. I very rarely see anyone misinterpret its meaning, and when I do, they are virtually always complete beginners to the language. The word "unreachable" in English literally describes "a location that cannot be reached", and that is precisely what the construct models.

> There are enough commits in Zig's codebase where developers made this mistake to disprove this.

I highly doubt that many -- if any -- of the examples you brought up were people misunderstanding the keyword. `unreachable` is not a complex construct, and is widely understood by basically everyone with a reasonable understanding of Zig. Instead, placing `unreachable` here is a simple logic error: the author thought this code path was unreachable, but it was not. Zig, including its standard library, is v0 software which undergoes rapid change -- there are going to be bugs. That isn't a reason to hinder the optimizer for everyone on the off-chance that someone deploys software without adequate testing in a security-critical environment in a manner where the failure happens to be caught by this mechanism (rather than being a more subtle logic error, or other illegal behavior).

**Comment by silversquirl** [comment]
> If you are not confident in the correctness of your assertions, you should not run the code in security-critical environments. 

I'd like to add that if you care about security you should be a) *exhaustively* testing *and* fuzzing your code, with runtime safety enabled; and possibly even b) shipping in `ReleaseSafe`, so that when something does go wrong, you immediately crash.

**Comment by andrewrk** [comment]
> Recent incorrect usages of `unreachable` in the Zig compiler

Your very first example is bogus. You're assuming the `catch unreachable` was an invalid assumption and the commit was to fix it but if you actually look at 0a3dff8125662feee7e348ed01478482ded0aa27 you'll see that's not the case.

Anyway it's moot. This proposal is never going to happen. It destroys the heart of the language.

It sounds like you're using ReleaseFast but you want to be using ReleaseSafe.
